staging/src/kubevirt.io/api/apitesting/roundtrip/compatibility.go

```diff
@@ -0,0 +1,446 @@
+)
+
+func (c *CompatibilityTestOptions) Complete(t *testing.T) *CompatibilityTestOptions {
+	t.Helper()
+
+	// Verify scheme
+	if c.Scheme == nil {
+		t.Fatal("scheme is required")
+	}
+
+	// Populate testdata dirs
+	if c.TestDataDir == "" {
+		c.TestDataDir = "testdata"
+	}
+	if c.TestDataDirCurrentVersion == "" {
+		c.TestDataDirCurrentVersion = filepath.Join(c.TestDataDir, "HEAD")
+	}
+	if c.TestDataDirsPreviousVersions == nil {
+		dirs, err := filepath.Glob(filepath.Join(c.TestDataDir, "release-*"))
+		if err != nil {
+			t.Fatal(err)
+		}
+		sort.Strings(dirs)
+		c.TestDataDirsPreviousVersions = dirs
+	}
+
+	// Populate kinds
+	if len(c.Kinds) == 0 {
+		gvks := []schema.GroupVersionKind{}
+		for gvk := range c.Scheme.AllKnownTypes() {
+			if gvk.Version == "" || gvk.Version == runtime.APIVersionInternal {
+				// only test external types
+				continue
+			}
+			if strings.HasSuffix(gvk.Kind, "List") {
+				// omit list types
+				continue
+			}
+			if ignoreCoreKinds.Has(gvk.Kind) {
+				// only test options types in the core API group
+				continue
+			}
+			if gvk == v1.VirtualMachineInstanceMigrationGroupVersionKind || gvk == v1.VirtualMachineInstancePresetGroupVersionKind || gvk == v1.VirtualMachineInstanceReplicaSetGroupVersionKind {
+				continue
+			}
+			gvks = append(gvks, gvk)
+		}
+		c.Kinds = gvks
+	}
+
+	// Sort kinds to get deterministic test order
+	sort.Slice(c.Kinds, func(i, j int) bool {
+		if c.Kinds[i].Group != c.Kinds[j].Group {
+			return c.Kinds[i].Group < c.Kinds[j].Group
+		}
+		if c.Kinds[i].Version != c.Kinds[j].Version {
+			return c.Kinds[i].Version < c.Kinds[j].Version
+		}
+		if c.Kinds[i].Kind != c.Kinds[j].Kind {
+			return c.Kinds[i].Kind < c.Kinds[j].Kind
+		}
+		return false
+	})
+
+	// Fill any missing objects
+	if c.FilledObjects == nil {
+		c.FilledObjects = map[schema.GroupVersionKind]runtime.Object{}
+	}
+	fillFuncs := defaultFillFuncs()
+	for k, v := range c.FillFuncs {
+		fillFuncs[k] = v
+	}
+	for _, gvk := range c.Kinds {
+		if _, ok := c.FilledObjects[gvk]; ok {
+			continue
+		}
+
+		obj, err := CompatibilityTestObject(c.Scheme, gvk, fillFuncs)
+		if err != nil {
+			t.Fatal(err)
+		}
+		c.FilledObjects[gvk] = obj
+	}
+
+	if c.JSON == nil {
+		c.JSON = json.NewSerializer(json.DefaultMetaFactory, c.Scheme, c.Scheme, true)
+	}
+	if c.YAML == nil {
+		c.YAML = json.NewYAMLSerializer(json.DefaultMetaFactory, c.Scheme, c.Scheme)
+	}
+
+	return c
+}
+
+func (c *CompatibilityTestOptions) Run(t *testing.T) {
+	usedHEADFixtures := sets.NewString()
+
+	var ranCurrentTests bool
+	for _, gvk := range c.Kinds {
+		t.Run(makeName(gvk), func(t *testing.T) {
+
+			t.Run("HEAD", func(t *testing.T) {
+				c.runCurrentVersionTest(t, gvk, usedHEADFixtures)
+				ranCurrentTests = true
+			})
+
+			for _, previousVersionDir := range c.TestDataDirsPreviousVersions {
+				t.Run(filepath.Base(previousVersionDir), func(t *testing.T) {
+					c.runPreviousVersionTest(t, gvk, previousVersionDir, nil)
+				})
+			}
```

Comment:
This basically says that we are trying to serialize `n-i` (where `i` is for now `1..3` ) to the `n-nth` version.

I do wonder if this is not excessive as we always support `n-1` to `n` upgrades.

Reply:
@xpivarc is the concern here cost or runtime of the test of something else?  

It seems like the tests are relatively cheap and quick, might not hurt to have it few releases back especially since these are v1 APIs

Reply:
@xpivarc Can you please take a look at the above comment.
 Is this going to be a merge blocker, if not it would be nice to get it merged before the upcoming release.

Reply:
This should be pretty fast, you are right. It is more about how we want to maintain this going forward and what kind of message this can give to users/vendors... As there is not much now we can take this as is but at the sig-api meeting I would like us  to talk about pros/cons and if any edge cases there might be... It would be also great to talk about how we extend this to validation (webhooks)...